internal/rpcserver/rpcserver.go

```diff
@@ -132,6 +132,13 @@ var (
 		Code:    dcrjson.ErrRPCNoWallet,
 		Message: "This implementation does not implement wallet commands",
 	}
+
+	// ErrExistsAddrIndexDisabled is an error returned to RPC clients when
+	// a method would need exists address indexing enabled but it is not.
+	ErrExistsAddrIndexDisabled = &dcrjson.RPCError{
```

Comment:
It appears that the intention here is that the interface funcs are expected to return this error, but that is not documented, and, I really don't think it should be structured that way anyway.

IMO, the handler in the RPC should be the one responsible for converting errors coming from the interface funcs into proper JSON-RPC errors (because the implementations will invariably be returning all types of errors).  You'll notice this is the way it's done throughout.

Reply:
I just wanted to keep the same error code for the same error, but if it is ok to return `dcrjson.ErrRPCInvalidParameter` for this as well, I would be glad to. It is an error that happens as a result of the operator of the node setting the config parameter to true, so maybe just as correct to call it an invalid parameter.

Reply:
Now only returning dcrjson.ErrRPCInvalidParameter. I think this is open to interpretation but it is true enough for me. I'm also open to alternatives.

I'm not sure that this doesn't need a version bump, but as the previous functioning was not documented, and it would be a very strange error code to listen for anyway as the default is to have exist address indexing turned on, I think it is ok.

Reply:
It seems like `dcrjson.ErrRPCInternal` would be the more appropriate error to use if `ExistsAddress` returns an error (even in the current code).  The current logic assumes that the err returned from `ExistsAddress` is due to a bad parameter when that may not be the case.  

Either way, as @JoeGruffins mentioned, this would change an error code that is being returned, so I'd have the same question as him as to whether this would need a version bump.

Reply:
Agree regarding `dcrjson.ErrRPCInternal`.  What I was trying to point out is not that the specific RPC error code is incorrect, rather that I would expect the RPC server handler itself to handle the conversion, because it is going to frequently happen that the errors coming out of the interfaces are NOT RPC errors.

I'm not against the notion of defining specific interface method semantics such that an implementer MUST return `rpcserver.ErrFoo` if `(condition)`.  This is done elsewhere in the code, so it's not out of character, but, in that case, the comment on the interface method(s) has to specifically define those semantics.

As it was written at the time of the comment in question, the RPC server handler just assumed the error coming back out of the call to `ExistsAddress` and `ExistsAddresses` was either going to be this special error or some other error, but the interface said nothing about it, so an implementer had no idea they had to do that, and the result would almost certainly be that eventually some implementer would return a general error that wouldn't be detected.

Anyway, it appears the latest iteration of this PR is no longer doing that, so it's now moot.

Reply:
Regarding the error code, I suppose to avoid the change altogether, the interface can pick up a new method `IsQueryable() bool` and the handler could keep semantics it does now where it first checks that and returns the same error code as now if it is not.

At that point, it definitely becomes an API error to call the `ExistsAddress` or `ExistsAddresses` methods, which would definitely set an error squarely in the internal camp.

However, I'm not really sure I like that solution as it feels a bit hacky to me.

I guess the most sane way to keep the current error codes is to do it the way @JoeGruffins had it originally where the interface methods are required return a specifically detectable error.

As mentioned above, I don't really have any issue with that approach *so long as the methods clearly call out the requirement*.

Reply:
In that case, what do you think about just having the interface return a `dcrjson.RPCError` and do the translation for both cases?  It feels awkward for it to sometimes return a specific `dcrjson.RPCError` error, and othertimes not, leading the caller to have to check for it

Reply:
That seems reasonable.  My only counter argument would be that it seems like it might be creeping RPC-related things (specifically what codes to return) back into the calling code, instead of it being something the RPC server handles and it would be pretty easy to change it without the RPC server getting bumped because it's not in the `rpcserver` code.

For those reasons, I think I would personally prefer the main code to know nothing at all about RPC error codes, however, I am happy to go with that approach is you guys prefer it.

Reply:
Ya, I agree with that, my preference would also be to have no RPC-related code in there, but it seems better than having it return a `dcrjson.RPCError` for just one specific case.  I'd also be good with the solution of leaving it closer to how it is currently, with something like the `IsQueryable() bool` that you suggested.

This is more probably complicated than it needs to be since the case checking if the index is enabled is [returning dcrjson.ErrRPCInternal](https://github.com/decred/dcrd/blob/master/internal/rpcserver/rpcserver.go#L1439), but the case checking the err returned from `ExistsAddress` is [returning dcrjson.ErrRPCInvalidParameter](https://github.com/decred/dcrd/blob/master/internal/rpcserver/rpcserver.go#L1455).

It seems that returning `dcrjson.ErrRPCInvalidParameter` in the second case is making an assumption of an invalid parameter, which may not be true, and it would be more correct to return `dcrjson.ErrRPCInternal`.  So, I think the best solution would be to just change that, and then we don't need this additional logic.  I'll defer to you on whether you want to avoid that, but it would be good to know more regarding the process of when/how you decide to bump the rpcserver version just for future reference as well.

Reply:
I think that if we really want to keep the same error codes, I would like to add the `IsQueryable() bool` to the interface.

But, in the case of these two handlers, I really don't think we are giving useful information with the separate error code. It would trigger due to a user controlled setting and the default is for indexing to be on. I cannot think of a situation where this would be singled out.  So, I would prefer to just return one error code. `dcrjson.ErrRPCInternal` is also fine with me. Although this wouldn't be the case for all handlers, I really don't think the additional logic is worth it in these two handlers. This does leave the question of needing to up the version however.

I would be fine with adding `IsQueryable() bool` for now and another TODO as well.

Reply:
Was just looking at the `CFilter` related handlers and the surrounding logic is similar, and I am thinking of handling like this.

If we want to keep the error codes the same, instead of using the `IsQueryable() bool`, I think it would be simpler to just embed `*indexers.ExistsAddrIndex` into `rpcExistsAddresser`:
```
type rpcExistsAddresser struct {
    *indexers.ExistsAddrIndex
}
```
And then check for `nil` in the applicable rpcserver code, eg:
```
if s.cfg.ExistsAddresser == nil {
    return nil, rpcInternalError("Exists address index disabled",
        "Configuration")
}
```

Reply:
Ok. It feels a little wierd, but I guess it's not a lot different than it was before. The interface now calls out that it may be unusable:
```
// ExistsAddresser may be nil. The RPC server must check for the presense of an
// ExistsAddresser before calling methods associated with it.
```

Reply:
I still kinda think `IsQueryable() bool` would be safer. Then there would be no possibility of panic if misused. Let me know.